File: src/mndl.hs
data Cmplx = Cmplx {im :: Double, re :: Double}

cmplxAdd :: Cmplx -> Cmplx -> Cmplx
cmplxAdd c1 c2 = Cmplx i r
    where
        r = re c1 + re c2
        i = im c1 + im c2

cmplxMul :: Cmplx -> Cmplx -> Cmplx
cmplxMul c1 c2 = Cmplx i r
    where
        r = (re c1)*(re c2) - (im c1)*(im c2)
        i = (re c1)*(im c2) + (im c1)*(re c2)
 
cmplxPow :: Cmplx -> Int -> Cmplx 
cmplxPow c n
    | n == 0    = Cmplx 0 1
    | otherwise = (cmplxPow c (n-1)) `cmplxMul` c
    
cmplxAbs :: Cmplx -> Double
cmplxAbs c = sqrt ((re c)**2+(im c)**2)

mndl :: Cmplx -> Int -> Cmplx
mndl c i
    | i == 0    = c
    | otherwise = ((mndl c (i-1)) `cmplxPow` 2) `cmplxAdd` c
    
mndlChar :: Cmplx -> Char
mndlChar c
    | cmplxAbs (mndl c 30) < 1000   = 'O'
    | otherwise                     = ' '

mndlLine :: Double -> String
mndlLine i = (map mndlChar c) ++ "\n"
    where c = map (Cmplx i) [(-2),(-1.96875)..1]


mndlString :: String
mndlString = concat (map (mndlLine) [1,0.96875..(-1)])
       
main :: IO ()
main = do
    putStr mndlString

File: src/mndl.lua
function cmplxAdd(c1,c2)
    return {
        r = c1["r"]+c2["r"],
        i = c1["i"]+c2["i"]
    }
end

function cmplxMul(c1,c2)
    return {
        r = c1["r"]*c2["r"] - c1["i"]*c2["i"],
        i = c1["r"]*c2["i"] + c1["i"]*c2["r"]
    }
end

function cmplxPow(c,n)
    if n==0 then
        return {
            r = 1,
            i = 0
        }
    else
        return cmplxMul(cmplxPow(c,n-1),c)
    end
end
    
function cmplxAbs(c)
    return math.sqrt(c["r"]^2+c["i"]^2)
end
    
function mndl(c,i)
    if i==0 then
        return c
    else
        return cmplxAdd(cmplxPow(mndl(c,i-1),2),c)
    end
end

c = {
    r = -2,
    i = 1
}

while c["i"]>=-1 do
    c["r"] = -2
    while c["r"]<=1 do
        if cmplxAbs(mndl(c,30)) < 1000 then
            io.write("O")
        else
            io.write(" ")
        end
        c["r"] = c["r"] + 0.03125
    end
    io.write("\n")
    c["i"] = c["i"] - 0.03125
end
